### DISCUSSION TITLE
Custom parser with devalue

### DISCUSSION THREAD

**Discussion by rChaoz** [socketio/socket.io#4979] [Q&A] [answered]
Hi! I was transfering data with Socket.IO nice and chill until I realised it uses `JSON.stringify()` under the hood for data serialization. I would like to instead use [devalue](https://www.npmjs.com/package/devalue). I looked at the default parser implementation and it seems quite complex, handling binary data and all. How should I go about using devalue (with its `stringify` and `parse`)?

I've also looked at the [sample custom parser](https://socket.io/docs/v4/custom-parser/#implementing-your-own-parser) and it seems much simpler than the default parser. How much is lost when switching to it due to this?

Would it makes sense to submit a feature request for this? I think with devalue in general being a very popular serialization option, it would be a nice addition. Many existing libraries/frameworks that need serialization already use it as the default over `JSON.stringify`.

**Comment by darrachequesne** [comment]
That's an interesting idea :+1: 

Would you have time to create a parser based on `devalue`? We could create a dedicated repo under the `socketio` organization.

**Reply by rChaoz** [reply]
I definitely would like to! I'm curious on how it should look like. Initially just replace `JSON.parse` with `devalue.parse` (and the same for stringify). Next, there are some immediate questions:

1. `JSON.stringify` never throws. `devalue` throws if it encounters any non-POJO, such as class instances. How should this be handled?
2. `devalue` can't handle binary data, so should the default parser's binary data handling (with deconstruct & reconstruct) be kept as-is?
3. If the answer to (1) is to leave it up to the user (which most libraries that use devalue seem to do) and to (2) is to leave it as-is, what else other than replacing `JSON` with `devalue` should be done?

**Reply by darrachequesne** [reply]
Here we go: https://github.com/socketio/socket.io-devalue-parser

> `JSON.stringify` never throws. `devalue` throws if it encounters any non-POJO, such as class instances. How should this be handled?

I'd say we should leave it to the user, which can for example implement a `toJSON()` method on their class.

> `devalue` can't handle binary data, so should the default parser's binary data handling (with deconstruct & reconstruct) be kept as-is?

That's a good question. Maybe we should keep it simple, and not support binary for now?

> If the answer to (1) is to leave it up to the user (which most libraries that use devalue seem to do) and to (2) is to leave it as-is, what else other than replacing `JSON` with `devalue` should be done?

That's about it I guess.

**Comment by rChaoz** [comment]
Would it make sense for the devalue repo to just be a fork of the standard parser so it stays up to date and keeps binary support? Or should I take the non-binary parts from the default one and just make a PR with that? 
 
  On Tue, 19 Mar 2024 at 16:51, Damien ***@***.***> wrote:   


Here we go: https://github.com/socketio/socket.io-devalue-parser


JSON.stringify never throws. devalue throws if it encounters any non-POJO, such as class instances. How should this be handled?


I'd say we should leave it to the user, which can for example implement a toJSON() method on their class.


devalue can't handle binary data, so should the default parser's binary data handling (with deconstruct & reconstruct) be kept as-is?


That's a good question. Maybe we should keep it simple, and not support binary for now?


If the answer to (1) is to leave it up to the user (which most libraries that use devalue seem to do) and to (2) is to leave it as-is, what else other than replacing JSON with devalue should be done?


That's about it I guess.

**Reply by darrachequesne** [reply]
Sure, that's valid too :+1:

**Reply by rChaoz** [reply]
I try responding once in my life via email and this happens. At least it did hide the addresses. Never again.

Anyway, which variant do you prefer? Because if it's the fork one I think you have to delete & recreate the repo as a fork before I can PR to it.

**Reply by rChaoz** [reply]
Hmm, found an issue. It seems like `devalue` is has `"type": "module"` in its `package.json`, so the CommonJS build of the parser fails. Should I remove it for now or do you have any other suggestions?

**Reply by darrachequesne** [reply]
I guess you can remove the CommonJS build for now.

**Reply by rChaoz** [reply]
It's done, you can check [the PR](https://github.com/socketio/socket.io-devalue-parser/pull/1). All tests are passing on my end, so it would be great to enable actions on the fork repo to see if they're okay in the CI pipeline as well.

**Reply by darrachequesne** [reply, marked as answer]
Here we go! https://www.npmjs.com/package/@socket.io/devalue-parser
